enterprise/server/quota/quota_manager.go

```diff
@@ -467,70 +477,116 @@ func (qm *QuotaManager) ApplyBucket(ctx context.Context, req *qpb.ApplyBucketReq
+func (qm *QuotaManager) addBucket(ctx context.Context, namespace string, bucket *qpb.Bucket) error {
 	if err := validateBucket(bucket); err != nil {
-		return nil, status.InvalidArgumentErrorf("invalid add_bucket: %s", err)
+		return status.InvalidArgumentErrorf("invalid add_bucket: %s", err)
 	}
 	row := bucketToRow(namespace, bucket)
 
-	err := qm.env.GetDBHandle().DB(ctx).Create(&row).Error
-	if err != nil {
-		return nil, err
-	}
-	return &qpb.ModifyNamespaceResponse{}, nil
+	return qm.env.GetDBHandle().DB(ctx).Create(&row).Error
 }
 
-func (qm *QuotaManager) updateBucket(ctx context.Context, namespace string, bucket *qpb.Bucket) (*qpb.ModifyNamespaceResponse, error) {
+func (qm *QuotaManager) updateBucket(ctx context.Context, namespace string, bucket *qpb.Bucket) error {
 	if err := validateBucket(bucket); err != nil {
-		return nil, status.InvalidArgumentErrorf("invalid update_bucket: %s", err)
+		return status.InvalidArgumentErrorf("invalid update_bucket: %s", err)
 	}
 	bucketRow := bucketToRow(namespace, bucket)
 
 	db := qm.env.GetDBHandle().DB(ctx)
 	res := db.Model(bucketRow).Where("namespace = ? AND name= ?", bucketRow.Namespace, bucketRow.Name).Updates(bucketRow)
 
 	if res.Error != nil {
-		return nil, res.Error
+		return res.Error
 	}
 	if res.RowsAffected == 0 {
-		return nil, status.InvalidArgumentErrorf("bucket %s doesn't exist", bucket.GetName())
+		return status.InvalidArgumentErrorf("bucket %q doesn't exist", bucket.GetName())
 	}
-	return &qpb.ModifyNamespaceResponse{}, nil
+	return nil
 }
 
-func (qm *QuotaManager) removeBucket(ctx context.Context, namespace string, bucketName string) (*qpb.ModifyNamespaceResponse, error) {
+func (qm *QuotaManager) removeBucket(ctx context.Context, namespace string, bucketName string) error {
 	dbh := qm.env.GetDBHandle()
-	err := dbh.TransactionWithOptions(ctx, db.Opts().WithQueryName("remove_bucket"), func(tx *db.DB) error {
+	return dbh.TransactionWithOptions(ctx, db.Opts().WithQueryName("remove_bucket"), func(tx *db.DB) error {
 		if err := tx.Exec(`DELETE FROM QuotaGroups WHERE namespace = ? AND bucket_name = ?`, namespace, bucketName).Error; err != nil {
 			return err
 		}
 		return tx.Exec(`DELETE FROM QuotaBuckets WHERE namespace = ? AND name = ?`, namespace, bucketName).Error
 	})
+}
+
+func (qm *QuotaManager) reloadNamespaces() (map[string]*namespace, error) {
+	config, err := fetchConfigFromDB(qm.env)
 	if err != nil {
 		return nil, err
 	}
-	return &qpb.ModifyNamespaceResponse{}, nil
+	namespaces := make(map[string]*namespace)
+	for nsName, nsConfig := range config {
+		ns, err := qm.createNamespace(qm.env, nsName, nsConfig)
+		if err != nil {
+			return nil, err
+
+		}
+		namespaces[nsName] = ns
+	}
+	return namespaces, nil
+}
+
+func (qm *QuotaManager) listenForUpdates(ctx context.Context) {
+	subscriber := qm.ps.Subscribe(ctx, pubSubChannelName)
+	defer subscriber.Close()
+	pubsubChan := subscriber.Chan()
+	for {
+		<-pubsubChan
```

Comment:
if you're just reading this channel forever, you can range over it:

for msg := range pubsubChan {
  ... blah ...
}

Reply:
Done